dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/tei/query/QueryContext.java

```diff
@@ -0,0 +1,212 @@
+            // to normalize overlapping attributes
+            .distinct()
+            .map( this::getFieldWithAlias )
+            .forEach( builder::field );
+
+        // todo: add sort by field if needed
+        return builder.build();
+    }
+
+    private Renderable getFieldWithAlias( String field )
+    {
+        return Field.of( TEI_TEI_DEFAULT_ALIAS, () -> field, field );
+    }
+
+    private From getFrom()
+    {
+        From.FromBuilder builder = From.builder();
+        builder.tablesOrSubQuery( getMainTable() );
+
+        List<Pair<Renderable, Renderable>> joinTablesAndConditions = getJoinTablesAndConditions();
+
+        joinTablesAndConditions.forEach(
+            renderableRenderablePair -> {
+                builder.tablesOrSubQuery( renderableRenderablePair.getLeft() );
+                builder.joinCondition( renderableRenderablePair.getRight() );
+            } );
+        return builder.build();
+    }
+
+    public Renderable getMainTable()
+    {
+        return Table.ofStrings( getMainTableName(), TEI_TEI_DEFAULT_ALIAS );
+    }
+
+    private List<Pair<Renderable, Renderable>> getJoinTablesAndConditions()
+    {
+        // TODO: implement logic for join when needed, i.e.
+        // when sorting on a data element is present
+        return Collections.emptyList();
+    }
+
+    private Where getWhere()
+    {
+        Where.WhereBuilder builder = Where.builder();
+        // conditions on programs (is enrolled in program)
+        getTeiQueryParams().getCommonParams().getPrograms()
+            .forEach( program -> builder.condition( () -> TEI_TEI_DEFAULT_ALIAS + "." + program.getUid() ) );
+        // conditions on filters/dimensions
+        getTeiQueryParams().getCommonParams().getDimensionIdentifiers()
+            .stream()
+            .map( this::toConditions )
+            .map( OrCondition::of )
+            .forEach( builder::condition );
+
+        return builder.build();
+    }
+
+    private List<Renderable> toConditions(
+        List<DimensionIdentifier<Program, ProgramStage, DimensionParam>> dimensionIdentifiers )
+    {
+        return dimensionIdentifiers.stream()
+            .filter( dimensionIdentifier -> dimensionIdentifier.getDimension().hasRestrictions() )
+            .map( this::toCondition )
+            .collect( Collectors.toList() );
+    }
+
+    private Renderable toCondition( DimensionIdentifier<Program, ProgramStage, DimensionParam> dimensionIdentifier )
+    {
+        return OrCondition.of(
+            dimensionIdentifier.getDimension().getItems().stream()
+                .map( item -> EventDataValueCondition.of( dimensionIdentifier, item, this ) )
+                .collect( Collectors.toList() ) );
+    }
+
+    public int bindParamAndGetIndex( Object param )
+    {
+        return parameterManager.bindParamAndGetIndex( param );
+    }
+
+    private static class ParameterManager
+    {
+
+        private int parameterIndex = 1;
```

Comment:
Should we use AtomicInteger here?

Reply:
it's not accessed by concurrent threads, so not needed